File: JPP/l2/z3/GF.py
GF_P = 1234577


def mult_inv(value):
    p = GF_P
    a0 = value
    y = 0
    x = 1

    while value > 1 and p != 0:
        q = value // p

        t = p

        p = value % p
        value = t
        t = y

        y = x - q * y
        x = t

    if value != 1:
        raise ValueError(f"Nie istnieje odwrotność {a0} modulo {GF_P}")

    if x < 0:
        x += GF_P
    return x


class GF:
    def __init__(self, _val):
        self.val = _val % GF_P

    def __add__(self, other):
        return GF(self.val + other.val)

    def __sub__(self, other):
        return GF(self.val - other.val)

    def __mul__(self, other):
        return GF(self.val * other.val)

    def __truediv__(self, other):
        return GF(self.val * mult_inv(other.val))

    def __floordiv__(self, other):
        return self.val * mult_inv(other.val)

    def __iadd__(self, other):
        self.val = (self.val + other.val) % GF_P
        return self

    def __isub__(self, other):
        self.val = (self.val - other.val) % GF_P
        return self

    def __itruediv__(self, other):
        self.val = (self.val * mult_inv(other.val)) % GF_P
        return self

    def __ifloordiv__(self, other):
        self.val = (self.val * mult_inv(other.val)) % GF_P
        return self

    def __eq__(self, other):
        return self.val == other.val

    def __ne__(self, other):
        return self.val != other.val

    def __lt__(self, other):
        return self.val < other.val

    def __gt__(self, other):
        return self.val > other.val

    def __le__(self, other):
        return self.val <= other.val

    def __ge__(self, other):
        return self.val >= other.val

File: JPP/l2/z2/GF.java
package z2;

public class GF {
    private static final long GF_P = 1234577;
    public long value;

    public GF(long _value){
        this.value = _value % GF_P;
    } 

    public static GF add(GF a, GF b)
    {
        return new GF(a.value + b.value);
    } 

    public static GF sub(GF a, GF b)
    {
        return new GF(a.value - b.value);
    } 

    public static GF mul(GF a, GF b)
    {
        return new GF(a.value * b.value);
    }

    private long mult_inv()
    {
        long p, a0, y, x;
        long a = this.value;
        p = GF_P;
        a0 = a;
        y = 0;
        x = 1;

        while (a > 1 && p != 0) {
            long q = a / p;
            long t = p;
            p = a % p;
            a = t;
            t = y;
            y = x - q * y;
            x = t;
        }

        if (a != 1) {
            System.err.printf("Nie istnieje odworotność %d modulo %d\n", a0, GF_P);
        }

        if (x < 0) {
            x += GF_P;
        }
        return x;
    }

    public static GF div(GF a, GF b)
    {
        return new GF(a.value * b.mult_inv());
    }

    public void iadd(GF other)
    {
        this.value = (this.value + other.value) % GF_P;
    }

    public void isub(GF other)
    {
        this.value = (this.value - other.value) % GF_P;
    }

    public void imult(GF other)
    {
        this.value = (this.value * other.value) % GF_P;
    }

    public void idiv(GF other)
    {
        this.value = (this.value * other.mult_inv()) % GF_P;
    }

    public static boolean eq(GF a, GF b)
    {
        return a.value == b.value;
    }

    public static boolean neq(GF a, GF b)
    {
        return !(GF.eq(a, b));
    }

    public static boolean lt(GF a, GF b)
    {
        return a.value < b.value;
    }

    public static boolean gt(GF a, GF b)
    {
        return a.value > b.value;
    }

    public static boolean leq(GF a, GF b)
    {
        return a.value <= b.value;
    }

    public static boolean geq(GF a, GF b)
    {
        return a.value >= b.value;
    }
}